### DISCUSSION TITLE
C# Had a few questions regarding intraday technical analysis

### DISCUSSION THREAD

**Discussion by preintel** [DaveSkender/Stock.Indicators#590] [Q&A] [answered]
1) Is it possible to pass one minute data to get technical results (for example if I have 2 days 1 minute prior data 7am-7pm and half of today's data 7am-12pm) can I pass all this as a contiguous sample and get results where 7pm data is next to the following day's 7am data. 
2) What error/result will I get if I don't pass enough data points to calculate a technical analysis result and should I catch this as an exception or is there a better approach?
3) A number of technical results will be smaller than the original sample size, is there a smart way programmatically to overlay the technical results on the same graph x-axis as the sample (OHLCV chart), eg should I offset the x-axis on the technical results by the number of missing first data points (for example the SMA line will start a few data points to the right of the price chart but still overlay the price chart)?

Thanks for creating this library! Going to sift through some of the source code and samples to check it out more. =)

**Comment by DaveSkender** [comment, marked as answer]
> Is it possible to pass one minute data to get technical results (for example if I have 2 days 1 minute prior data 7am-7pm and half of today's data 7am-12pm) can I pass all this as a contiguous sample and get results where 7pm data is next to the following day's 7am data.

Yes, there's nothing wrong with providing contiguous data.  Dates are not actually used in any of the formulas, but are rather kept together with results for context.  The only word of caution here is that the 7pm to 7am gap could be missing after-hours information that might be useful; you'd have to interpret a potential rapid change in prices accordingly.

> What error/result will I get if I don't pass enough data points to calculate a technical analysis result and should I catch this as an exception or is there a better approach?

If you do not provide enough historical quotes for calculation, you'll get (and can catch) a `BadQuotesException`.  How you handle it is up to you: here’s a few ideas: 1) go get more quotes and try again, 2) elevate the error to your calling process or to a user somehow, or 3) return an empty set (ghost it 👻).

Generally speaking, I recommend providing more than the minimum required history for any indicator that has a ⌛  _Convergence Warning_ in the documentation because there are scenarios where precision improves when more history is provided.  In other words, if you're always only providing minimum history, there's a good chance you're using less precise indicator results.

> A number of technical results will be smaller than the original sample size, is there a smart way programmatically to overlay the technical results on the same graph x-axis as the sample (OHLCV chart), eg should I offset the x-axis on the technical results by the number of missing first data points (for example the SMA line will start a few data points to the right of the price chart but still overlay the price chart)?

By design and as the default, almost all the indicators in this library return the same number of elements as the provided historical quotes -- they're exactly the same length, though with `null` values for the incalculable warmup periods.  We've done this specifically for that reason.  See docs for info on which ones do not (e.g. Renko Chart).

Inversely, to exclude this initialization period, you can opt-out with the `.RemoveWarmupPeriods()` helper function to prune the recommended warmup and convergence periods.

**Reply by preintel** [reply]
Thanks I got it working but now trying to determine the best way to save the output. If I am storing my quote data in a dictionary with datetime keys I was wondering if all the technical indicators from Skender outputs a datetime variable so I can simply add each technical output to the same keys of my original quote data for easier accessibility? I see that SMA and RSI have a datetime property do the other technical outputs all return this property?

Basically every minute I get a new quote added to my source quote data which I need to call each technical analysis method in Skender again. Since the output is IEnumerable<Type> but the dateTime property is within the IEnumerable itself I end up having to loop through each item in the IEnumerable or convert it to a list/array then loop, I figure there is no way to take a given DateTime and access the corresponding values in the IEnumerable like IEnumerable<RsiResults>[DateTime]) 
- In this scenario should I call each technical analysis function every minute (as my source quote data is updated), loop through each technical IEnumerable result output and manually store the values into a Dictionary/class structure with DateTime keys that I can easily index for future use or is there a better approach to storing and accessing the results? Not sure how much memory/processing this will take for larger datasets if it gets called every minute...

**Reply by DaveSkender** [reply]
> [do] all the technical indicators from Skender outputs a datetime variable so I can simply add each technical output to the same keys of my original quote data for easier accessibility?

Yes, the library always includes the matching `Date` information for indicators.  The `results.Find(lookupDate)` helper function might be useful, depending on what you’re doing; though there are probably more efficient approaches for bulk operations.

Under the hood, this helper function is doing this:
```csharp
return results.FirstOrDefault(x => x.Date == lookupDate);
```

**Reply by DaveSkender** [reply]
On a related note, I have some thought on incremental calculations that I’m jotting down in the roadmap - see #396

**Reply by preintel** [reply]
Thank you for the response. It seems like you have built a trading program already and I am starting to make one and wondering if you could give me any pointers on algotrading using this library, seems like the market is quite random and hard to predict even with all this technical information or maybe I am just picking the wrong stocks?

**Reply by DaveSkender** [reply]
Here’s a bit of my philosophy, if that helps: [#277](https://github.com/DaveSkender/Stock.Indicators/discussions/277#discussioncomment-287215).

**Reply by preintel** [reply]
Thanks I will read up more on what technical functions to use. 
I was still looking for a way to reduce the amount of CPU usage by not calling the technical analysis functions every minute when a single candlestick is added to my source data, does the TA library support  methods which use previously provided data and simply add a single OHLCV to the end of the array or will I have to call the function each time I get a new data point to get accurate results?

**Reply by DaveSkender** [reply]
I’ll be adding incremental calculations in v2 where possible.  See #396 for more info.
